src/mpid/ch4/src/ch4_impl.h

```diff
@@ -1146,4 +1147,40 @@ MPL_STATIC_INLINE_PREFIX MPI_Op MPIDIU_win_acc_get_op(int index)
         return MPIR_Op_builtin_get_op(index);
     }
 }
+
+MPL_STATIC_INLINE_PREFIX bool MPIDIG_rma_recv_am_always_poll(void)
+{
+    bool always_poll = false;
+    int interval;
+
+    MPIR_cc_incr(&MPIDIG_global.rma_recv_am_poll_cntr, &interval);
+
+    if (MPIR_CVAR_CH4_RMA_RECV_AM_PROGRESS_INTERVAL > 0) {
+        /* User explicitly controls the polling frequency */
+        always_poll =
+            ((interval + 1) % MPIR_CVAR_CH4_RMA_RECV_AM_PROGRESS_INTERVAL == 0) ? true : false;
+    } else if (MPIR_CVAR_CH4_RMA_RECV_AM_PROGRESS_INTERVAL == -1) {
+        /* Runtime control is on.
+         * - Always poll if RMA receive AM is active
+         * - Otherwise poll with longer interval */
+        if (MPL_atomic_load_int(&MPIDIG_global.rma_recv_am_active)) {
+            always_poll = true;
+        } else {
+            always_poll =
+                ((interval + 1) % MPIR_CVAR_CH4_RMA_RECV_AM_PROGRESS_INTERVAL_LONG ==
+                 0) ? true : false;
```

Comment:
@minsii This branch clearly says the semantics is not *always*, so I would suggest remove the "always" (in addition of the "recv") from the function name, simply call it `MPIDIG_rma_need_poll_am()`.

Reply:
Agree to remove `always`. But I do want to keep the meaning that this function defines only the progress demand for incoming RMA AM. E.g., even this function returns false, we still need check outgoing RMA AM in flush calls and decide to poll progress if it found any outgoing RMA AM.

Reply:
If you really insist, how about `MPIDIG_rma_need_poll_am_recv()`? I don't really like having the extra `recv` suffix, but since I got the `need_poll_am` part, I think I can tolerate the suffix. @raffenet What do you think?

Reply:
You made the point that `am_recv` or `recv_am` might be unclear as it also includes receiving the ack on the origin process (which we don't care here). I agree with the point. So, how about `MPIDIG_rma_target_need_poll_am` or `MPIDIG_rma_need_poll_target_am`?

Reply:
>  how about `MPIDIG_rma_target_need_poll_am` or `MPIDIG_rma_need_poll_target_am`?

Well, when we read this name, the natural question is -- "which target"? That's why I think the word "target" on its own is very confusing.

Reply:
Target is the standard term to describe the process who receives RMA operation in a RMA communication.

Reply:
The confusing part is not what "target" means in the context of RMA. The confusing part is target didn't say anything about which origin, which target, or which process. There are plenty of variables names having "target" but often means other processes. Here you want it to mean the opposite, so it is confusing.

Reply:
How about "non-origin"?

Reply:
Back up a little, the full sentence your wanted to say here is `MPIDIG_rma_need_poll_am_because_of_[this_reason]` and we are discussing a good word to describe this reason. But even we have that word, if your take out the "because_of" part, the phrase is very confusing: `MPIDIG_rma_need_poll_am_reason`? 

I guess I am back to insist on just `MPIDIG_rma_need_poll_am` since that is complete on its own.

Reply:
We specifically use `origin_cbs` and `target_msg_cbs` to distinguish the AM received on origin process (ACK) and that received on target process in our code. The new flag is set only when any `target_msg_cbs` is called, and the function decides the polling strategy only for progressing `target_msg_cbs`. Thus, I need have the meaning in the name.

I think we cannot make conclusion for the naming. Thus, hope @raffenet can raise the third voice so we can finish the PR.

Reply:
Of the names suggested in this thread, I prefer `MPIDIG_rma_need_poll_am`. I don't think the reason needs to be stated in the API. The function is there to let us know if we need to poll (progress?) active messages for RMA. Why we need to poll is an implementation detail.

Reply:
Alright. Will make changes based on the naming of `MPIDIG_rma_need_poll_am`, although I still think it is confusing as we need check additional conditions outside of the function.

Reply:
> I still think it is confusing as we need check additional conditions outside of the function.

I see. If you don't like it because the function is incomplete, then why don't you roll the other condition into this function as well?

Reply:
We need check local completion for flush_local but remote completion for flush. I don't like mix the two cases in a single function and pass a parameter to distinguish them. IMO, they should belong to the original functions.

Reply:
Do we need poll am in `flush_local` when there is no outstanding local operations?

Reply:
I don't see why we need skip poll am in `flush_local`. We want to be consistent for all flush calls.

Reply:
The purpose of MPI_Win_flush_local is to be able to reuse the local buffer. So semantically user won't expect this function to help any remote progress.

It is not a strong reason, since MPI_Win_flush doesn't guarantee remote progress either (although some user may expect that). But it is the same reason that we decide not to call general progress in, for example, `MPI_Put`. I think narrow down our scope to `MPI_Win_flush` is good for now.

Reply:
User would expect some level of progress in flush_local. E.g., such loop may exist in user program and low-frequency progress is good to have.
```while(a<100) {fop(a/*result buf*/); flush_local(); /*complete atomic get*/}```

Reply:
The MPI standard only mentions `put`, `get`, and `accumulate` functions for `MPI_Win_flush_local`, though.

But as I relented, there isn't a strong reason that we choose to poke additional am progress in `MPI_Win_flush` vs poking progress in any other MPI functions such as `MPI_Put`. In a passive window, there isn't any semantics to guarantee progress. We choose `MPI_Win_flush` simply as a compromise, because that is a function that user would expect to be most expensive (other than those epoch closing syncs, which we poll am progress always). If you want to include `MPI_Win_flush_local`, it is your call. Then of course, your code and semantics gets a bit more complex. I'll leave to your judgement.

Reply:
Please trust me that many places in the standard mention `accumulate`  which actually means all accumulate-like operations, including acc, get_acc, fop, cswap. (I know it is confusing).

I still think it is straightforward and reasonable to also add the progress opt in `flush_local`. Dropping it will not make the semantics of `MPIDIG_rma_need_poll_am` more clear. Also, moving the origin-issued-am check into `MPIDIG_rma_need_poll_am` only for `flush_{all}` will make the code inconsistent. 

Anyway, I will rename the function/flag/CVAR, but let's leave the code logic as is.